app/components/Article/Contents.tsx

```diff
@@ -96,42 +96,57 @@ const normalizeForComparison = (text: string): string => {
     .replace(/[‒–—]/g, '-') // Replace figure-dash, en-dash and em-dash with hyphen
 }
 
-const glossaryInjecter = (pageid: string, glossary: Glossary) => {
-  const seen = new Set()
+const glossaryInjecter = (pageid: string, glossary: Glossary, seen: Set<string>) => {
   return (html: string) => {
-    // Normalize for comparison to find terms regardless of formatting
-    const normalizedHtml = normalizeForComparison(html)
-
-    return Object.values(glossary)
+    const sortedGlossary = Object.values(glossary)
       .filter((item) => item.pageid != pageid)
       .sort((a, b) => (b.alias?.length ?? 0) - (a.alias?.length ?? 0))
-      .reduce((currentHtml, {term, alias}) => {
-        // Create a normalized pattern for matching variations in formatting
-        const normalizedAlias = normalizeForComparison(alias || '')
-        const match = new RegExp(`(^|[^\\w-])(${normalizedAlias})($|[^\\w-])`, 'i')
-
-        // Check if the term exists in the normalized HTML
-        if (!seen.has(term) && normalizedHtml.search(match) >= 0) {
-          seen.add(term)
-
-          return currentHtml.replace(match, '$1<span class="glossary-entry">$2</span>$3')
-        }
-        return currentHtml
-      }, html)
+
+    // Track replacements with placeholders to avoid overlapping matches
+    const replacements: Array<{placeholder: string; replacement: string; term: string}> = []
+    let result = html
+    let placeholderCounter = 0
+
+    // Replace ALL occurrences of each term with placeholders
+    // This prevents shorter terms from matching within longer terms
+    for (const {term, alias} of sortedGlossary) {
+      const normalizedAlias = normalizeForComparison(alias || '')
+      const pattern = new RegExp(`(^|[^\\w-])(${normalizedAlias})($|[^\\w-])`, 'gi')
+
+      result = result.replace(pattern, (match, prefix, matchedTerm, suffix) => {
+        const placeholder = `__GLOSSARY_${placeholderCounter++}__`
```

Comment:
(minor) if there is only 1 loop and not doing replacements.push first and replacements.forEach second, then we wouldn't need even placeholder, placeholderCounter and just do the regex replace directly with the span (..but not a big deal if current code works fine)